clients/rospy/src/rospy/impl/tcpros_pubsub.py

```diff
@@ -386,15 +386,17 @@ def __init__(self, connection, queue_size):
 
         self._lock = threading.Lock()
         self._cond_data_available = threading.Condition(self._lock)
-        self._connection.set_cleanup_callback(self._closed_connection_callback)
         self._queue = []
         self._error = None
 
         self._thread = threading.Thread(target=self._run)
         self._thread.start()
 
+        self._connection.add_cleanup_callback(self._closed_connection_callback)
```

Comment:
That's just where it ended up when I added it back, but it doesn't matter.  Here's how I understand it.

There is potentially a race here either way, but slightly less with something following up the thread start.  I'd love to find a cleaner way to make sure that `QueuedConnection` acquires the `Lock` before the callback does, because if it doesn't then it will never be notified.  It is *probably* never going to be an issue because of what follows this constructor.

Starting from `_PublisherImpl.add_connection()`, the `QueuedConnection` wrapper is created before the critical section of `_TopicImpl.add_connection()` which uses the same `Lock` as `_TopicImpl.close()` (which is what closes the  the cleanup callback).  So, there's no way the callback can happen before this constructor is off the stack and the critical section of `_TopicImpl.add_connection()` will *probably* not finish before `QueuedConnection` can make it to the `Lock`.